### PR TITLE
Add a system overview section on the landing page #867

### LINKED ISSUE #867: general overview of tox - documentation
Hi, 
I would like to understand what is the basic flow of tox (which is an amazing project by the way !):
- I understand there is a run of setup.py in both environments (2.7, 3.6).
- But then, regarding the tests ran with pytest (or other test suite): are they running against package in source code, or against package installed in site-packages (by setup.py install) ? 

It's important in my case because my setup.py downloads a binary that I need and puts it in the site-packages directory, which is not originally in my source code package (because I need it to be cross-platform).

Thanks for help !

### PR DESCRIPTION
**Pull request by gaborbernat**
I find we lack a quick two-minute high-level overview of what tox is. This should help new people quickly understand easier what tox actually does and expose common gotchas.

See output at https://tox-gaborbernat.readthedocs.io/en/general/#system-overview.

Resolves #867.

### CODE REVIEW

**Comment by obestwalter** [comment]
Oh good, this would also address where we started discussing these things: https://github.com/tox-dev/tox/issues/478.

What should be covered though in the explanation is, that the visualized default flow is highly configurable (e.g. you could not build a package and you could also switch off installing the package (and use tox in a `setup.py` less project) or you could install a devlopment version. 

I have no idea yet how and where this should be visualized/documented, but I think it should definitely be part of a high level overview to make clear that tox can do much more than the standard flow.

**Comment by gaborbernat** [comment]
I know, and if you read the explanation for the phases you can see I pointed out clearly where things change behaviour.

**Comment by obestwalter** [comment]
I have to admit that I did not read very carefully on the first go, I will have a closer look tonight and might have some ideas for improvments.

**Comment by obestwalter** [comment]
Hi @gaborbernat, I will be able to dedicate next friday to tox properly and try to catch up then.

**Comment by obestwalter** [comment]
BTW - did you watch the video? What do you think about us trying to apply this to the tox documentation?

**Comment by gaborbernat** [comment]
I did not watch the video but was there in the audience. It makes some good points. As far as our project goes I think we just need to have a different page for the plugins.

**Comment by obestwalter** [comment]
I think I will take the friday to have a proper look at our existing docs and go from there.

File: doc/index.rst [hunk 1]
```diff
@@ -55,6 +55,72 @@ you'll note that it runs much faster because it keeps track of virtualenv detail
...
+
+3. **environment** - for each tox environment (e.g. ``py27``, ``py36``) do:
+
+    1. **environment creation**: create a fresh environment, by default Python virtual environments
+    are used via the virtualenv_ project. tox will automatically
+    try to discover a valid Python interpreter version by using the environment name (e.g. ``py27``
+    means Python 2.7 and the ``basepython`` configuration value) and the current operating system
+    ``PATH`` value. This step is done at first run only (or when :confval:`basepython` value
+    changes) and re-used at subsequent runs, unless the user requests explicitly recreating this
+    by passing in the ``-r`` flag.
```
Comment (obestwalter) [inline doc/index.rst]: A bit more handwavy but maybe sufficient here?

    This is created at first run only to be re-used at subsequent runs. If certain aspects of the project change, a re-creation of the environment is automatically triggered. To force the recreation (e.g. when changes weren't properly picked up), tox can be invoked with ``-r``/``--recreate``.

File: doc/index.rst [hunk 2]
```diff
@@ -55,6 +55,72 @@ you'll note that it runs much faster because it keeps track of virtualenv detail
...
+
+    2. **install** (optional): install the environment dependencies specified inside the
+    :confval:`deps` configuration section, and then the earlier packaged source distribution.
+    By default ``pip`` is used to install packages, however one can customise this via
+    :confval:`install_command`. Note ``pip`` will not update project dependencies (specified either
+    in the ``install_requires`` or the ``extras`` section of the ``setup.py``) if any version already
+    exists in the virtual environment; therefore we recommend to recreate your environments entirely
```
Comment (obestwalter) [inline doc/index.rst]: "entirely" is not necessary here (and maybe even potentially confusing?), as there is no way to recreate the environment partially.

File: doc/index.rst [hunk 3]
```diff
@@ -55,6 +55,72 @@ you'll note that it runs much faster because it keeps track of virtualenv detail
...
+    whenever your project dependencies change.
+
+    3. **commands**: run one by one the specified commands. Whenever the exit code of any of them
```
Comment (obestwalter) [inline doc/index.rst]: How about:

    run the specified commands in the specified order.

File: doc/index.rst [hunk 4]
```diff
@@ -55,6 +55,72 @@ you'll note that it runs much faster because it keeps track of virtualenv detail
...
+    is not zero stop, and mark the environment failed. Note, starting a command with a single dash
+    character means ignore exit code.
```
Comment (obestwalter) [inline doc/index.rst]: Or:

   To ignore the exit code of a command, a dash can be prefixed to the command.

File: doc/index.rst [hunk 5]
```diff
@@ -55,6 +55,72 @@ you'll note that it runs much faster because it keeps track of virtualenv detail
...
+6. **report** print out a report of outcomes for each tox environment:
+
+   .. code:: bash
+
+      ____________________ summary ____________________
+      py27: commands succeeded
+      ERROR:   py36: commands failed
+
+    Only if all environments finished with success will the exit code of tox be success. In this
+    case you'll also see the message ``congratulations :)``.
```
Comment (obestwalter) [inline doc/index.rst]: or:

    Only if all environments ran successfully tox will return `0` (success). In this case you'll also see the message ``congratulations :)``.

File: doc/index.rst [hunk 6]
```diff
@@ -55,6 +55,72 @@ you'll note that it runs much faster because it keeps track of virtualenv detail
...
+
+tox will take care of environment isolation for you: it will strip away all operating system
+environment variables not specified via :confval:`passenv`. Furthermore, it will also alter the
+``PATH`` variable so that your commands resolve first and foremost within the current active
+tox environment. It's possible to fallback to executables outside of this, however these need to
+be explicitly enabled via the :confval:`whitelist_external` configuration variable.
```
Comment (obestwalter) [inline doc/index.rst]: This is not quite accurate: tox uses whatever is available, but will warn when it was not expllicitly whitelisted.

How about:

    In general all executables in the path are available in ``commands``, but tox will emit a warning if it was not explicitly allowed via :confval:`whitelist_external`.

File: doc/index.rst [hunk 7]
```diff
@@ -55,6 +55,72 @@ you'll note that it runs much faster because it keeps track of virtualenv detail
...
+
+      python setup.py sdist
+
+   Note that for this operation the same Python environment will be used as the one tox is
+   installed into (therefore you need to make sure that it contains your build dependencies).
+   Skip this step for application projects that don't have a ``setup.py``.
+
+3. **environment** - for each tox environment (e.g. ``py27``, ``py36``) do:
+
+    1. **environment creation**: create a fresh environment, by default Python virtual environments
```
[the remark this answers is not in the document]
Comment (gaborbernat) [reply doc/index.rst]: I prefer having it as phase names, to help new users grasp tox run concepts. Hence I still think **environment creation** is better.

File: doc/index.rst (hunk 1 above)
Comment (gaborbernat) [reply doc/index.rst]: I'll use most of it other than the `` (e.g. when changes weren't properly picked up)``. Some changes are not tracked, but I don't think that warrants it being stated not properly.

File: doc/index.rst (hunk 7 above)
[the remark this answers is not in the document]
Comment (obestwalter) [reply doc/index.rst]: Sorry if that wasn't clear - I wasn't referring to the phase name, but to the explanation after that. Basically to shorten the first sentence and add the replacement info via plugin there as well (but I am not sure, if that is not overloading that part).

**Pull request merged**